### ISSUE TITLE
Support checking `...` args, Dots(), Lazydots() 

### ISSUE THREAD

**Issue by moodymudskipper** [moodymudskipper/typed#16]
would require that all dot args satisfy to the condition

OR

We might have a special typechecker `Dots()` which might be more flexible, it could work like `List()` on `list(...)`, or maybe substituted dots. 

In particular it would be very easy to check dots length or names, we might have an argument `each` to apply assertiton on each so `... ? Integer()` and `... ? Dots(each = Integer())` are equivalent. 

We might also have `Lazydots()` to, which would work like `List()` of `eval(substitute(alist(...)))` (related to `Lazy()`).

**Comment by moodymudskipper** [comment]
`Dots()` is very similar to `List()`, and `List()` would need this `each` argument too the only thing is that `function(... =? List())` would mean that each dot arg is a list.

So we could have `function(... =? Dots(3))` translated to `check_arg(list(...), List(3))` OR have `Dots()` be an alias to `List()` unless detected in arg, and then we can keep  `check_arg(list(...), Dots(3))`, I currently don't like the latter too much.

Lazy dots would work too : 

`function(... =?~ Dots(each = Symbol())` would be translated to : `check_arg(eval(substitute(alist(...)), List(each = Symbol())`

or `check_arg(substitute(...()), Pairlist(each = Symbol())`

`function(... =? Character())`  should be translated to `check_arg(list(...), List(each = Character()))`

**Comment by moodymudskipper** [comment]
`each` is implemented for `List`, `Pairlist` and `Data.frame`, it seems to work quite well : 

``` r
library(typed, warn.conflicts = FALSE)
test <- function() {
  List(each=Character()) ? x
  x <- list(1,2)
}
test()
#> Error: In `test()` at `x <- ...`:
#> column 1 type mismatch
#> `typeof(value)`: "double"   
#> `expected`:      "character"

test <- ? function(x= ?Data.frame(each=Double())) {
  x
}
test(iris)
#> Error: In `test(iris)` at `check_arg(x, Data.frame(each = Double()))`:
#> wrong argument to function, column 5 ("Species") type mismatch
#> `typeof(value)`: "integer"
#> `expected`:      "double"
```

<sup>Created on 2020-11-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by moodymudskipper** [comment]
This is implemented in 75e6f3260555b8708f0beb475b46e8a122dc6830

``` r
library(typed, warn.conflicts = FALSE)

test <- ? function(...= ? Double()) {
  "hello"
}
test(1, 2)
#> [1] "hello"
test(1,"a")
#> Error: In `test(1, "a")` at `check_arg(list(...), List(each = Double()))`:
#> wrong argument to function, element 2 type mismatch
#> `typeof(value)`: "character"
#> `expected`:      "double"

test <- ? function(...= ?~ Symbol()) {
  "hello"
}
a <- 1
test(a, a)
#> [1] "hello"
test(a, 1)
#> Error: In `test(a, 1)` at `check_arg(substitute(...()), Pairlist(each = Symbol()))`:
#> wrong argument to function, element 2 type mismatch
#> `typeof(value)`: "double"
#> `expected`:      "symbol"

test <- ? function(...= ? Dots(2)) {
  "hello"
}
test(1, 2)
#> [1] "hello"
test(1)
#> Error: In `test(1)` at `check_arg(list(...), Dots(2))`:
#> wrong argument to function, type mismatch
#> `length(value)`: 1
#>      `expected`: 2

test <- ? function(...= ?~ Dots(2,,~is.symbol(.[[1]]) && is.double(.[[2]]))) {
  "hello"
}
test(a, 1)
#> [1] "hello"
test(a, a)
#> Error: In `test(a, a)` at `check_arg(substitute(...()), Dots(2, , ~is.symbol(.[[1]]) && is.double(.[[2]])))`:
#> wrong argument to function, mismatch
#> `is.symbol(value[[1]]) && is.double(value[[2]])`: FALSE
#> `expected`:                                       TRUE
```

<sup>Created on 2020-11-03 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>
